webapp/openapi_parser.py

```diff
@@ -0,0 +1,37 @@
+from collections import defaultdict
+from yaml import load
+from yaml import Loader
+
+
+def parse_openapi(definition):
+    """
+    Takes an OpenAPI definition in YAML and returns it as a dictionary, with
+    endpoints grouped by tags.
+
+      Parameters:
+        definition: The OpenAPI YAML defintion (loaded from file or URL)
+
+      Returns:
+        dict: A dictionary of tags, each tag containing a list of API
+        endpoints (as dicts)
+    """
+    loaded_definition = load(definition, Loader)
+
+    tagged_definition = defaultdict(list)
+
+    for endpoint, methods in loaded_definition["paths"].items():
+        for method, definition in methods.items():
+            if method != "parameters":
```

Comment:
hmm looking at this more closely now it's cleaner...

1) why are we skipping over `parameters` ? Can you add a comment to explain
2) inside the loop of different methods, what we actually accumulate is a mapping of `{endpoint: methods}` ... so we're going to be adding the whole section for each method on a given endpoint (unless it only has parameters?) - that's why the simple version gave you duplicates.

Reply:
I'll add a comment to explain, but the gist of it is that in the YAML definition, "parameters" are listed on the same level as methods. So we can't loop through an endpoint and assume that every item we encounter is a method structured in a recognisable way, we have to have a separate case to account for parameters that affect the endpoint as a whole. In essence it may be better to refactor the code to loop through "attributes" instead of "methods" and then have a case for `if attribute == "parameters"`

Reply:
On the second point - the tags are used to group each endpoint in the usual Swagger UI, but these tags are stored *per method*, not per endpoint, even though every method under an endpoint has the same tag. So we need the complex logic to verify that when a method is checked for its tags, the endpoint that it belongs to hasn't already been added to the tagged dictionary of endpoint lists